### PR TITLE
add initial object logging 

### PR DESCRIPTION
**Pull request by jerryjliu**
- add very basic initial logger abstraction 
- store intermediate outputs during create/refine specifically 
- TODO: add for tree summarize, add for selecting nodes, add for building the index

### CODE REVIEW

**Comment by brunoramirez12** [comment]
documentati
[base.txt](https://github.com/jerryjliu/llama_index/files/10983478/base.txt)
on for base.py

File: gpt_index/logger/base.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    def reset(self) -> None:
+        """Reset logs."""
+        self._logs = []
+
+    def set_metadata(self, metadata: Dict) -> None:
+        """Set metadata."""
+        self._metadata.update(metadata)
+
+    def unset_metadata(self, metadata_keys: Dict) -> None:
```
Comment (MatthewCochrane) [inline gpt_index/logger/base.py:24]: Is there a reason for metadata_keys to be a `dict` here rather than a `set` or `list`?

Comment (jerryjliu) [reply gpt_index/logger/base.py:24]: oh they should be a set :)

Comment (jerryjliu) [reply gpt_index/logger/base.py:24]: good catch!

**Pull request merged**